File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/usecases/TopUpConfirmingTest.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.commands.ConfirmTopUp;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import it.giannibombelli.wsc2026.giftcard.infrastructure.SqliteGiftCardRepository;
import it.giannibombelli.wsc2026.testsupport.AggregateFactory;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class TopUpConfirmingTest {
    private GiftCardRepository repository;
    private TopUpConfirming confirming;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("giftcard");
        repository = new SqliteGiftCardRepository(dataSource);
        confirming = new TopUpConfirming(repository);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new TopUpConfirming(null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullCommand() {
            assertThatThrownBy(() -> confirming.invoke(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldPersistTopUpConfirmation() {
            GiftCard giftCard = AggregateFactory.getSavedGiftCard(repository);
            Money topUpAmount = new Money(new BigDecimal("33.00"));

            confirming.invoke(new ConfirmTopUp(giftCard.id(), topUpAmount));

            GiftCard persisted = repository.findById(giftCard.id()).orElseThrow();
            assertThat(persisted.balance()).isEqualTo(topUpAmount);
        }

        @Test
        void shouldFailIfCardDoesNotExist() {
            GiftCardId nonExisting = EntityId.generate(GiftCardId::new);
            Money amount = new Money(new BigDecimal("10.00"));

            assertThatThrownBy(() -> confirming.invoke(new ConfirmTopUp(nonExisting, amount)))
                .isInstanceOf(IllegalStateException.class);
        }
    }
}

File: typescript/tests/giftcard/application/usecases/topUpConfirming.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { ConfirmTopUp, confirmTopUp } from '@/giftcard/application/commands/confirmTopUp.js';
import { TopUpConfirming } from '@/giftcard/application/usecases/topUpConfirming.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { SqliteGiftCardRepository } from '@/giftcard/infrastructure/sqliteGiftCardRepository.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';
import { GiftCardAggregateFactory } from '../../../testsupport/giftcard/aggregateFactory.js';

describe('TopUpConfirming', () => {
  let repository: SqliteGiftCardRepository;
  let confirming: TopUpConfirming;

  beforeAll(() => {
    const database = DatabaseSetup.initializeInMemoryDb('giftcard');
    repository = new SqliteGiftCardRepository(database);
    confirming = new TopUpConfirming(repository);
  });

  describe('construction', () => {
    it('rejects null repository', () => {
      expect(() => new TopUpConfirming(null as unknown as SqliteGiftCardRepository)).toThrow();
    });
  });

  describe('execution', () => {
    it('rejects null command', () => {
      expect(() => confirming.invoke(null as unknown as ConfirmTopUp)).toThrow();
    });

    it('should persist top-up confirmation', () => {
      const giftCard = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const topUpAmount = new Money(33);

      confirming.invoke(confirmTopUp(giftCard.id(), topUpAmount));

      const persisted = repository.findById(giftCard.id());
      expect(persisted).not.toBeNull();
      expect(persisted!.balance()).toEqual(topUpAmount);
    });

    it('should fail if card does not exist', () => {
      const nonExisting = generateId((value) => new GiftCardId(value));
      const amount = new Money(10);

      expect(() => confirming.invoke(confirmTopUp(nonExisting, amount))).toThrow();
    });
  });
});
